### PR TITLE
[GEOT-6377]: added new filter function arrayAnyMatch and a PostGIS en…

### PR DESCRIPTION
**Pull request by mbarto**
…coding using its ANY function

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+package org.geotools.filter.function;
+
+import static org.geotools.filter.capability.FunctionNameImpl.parameter;
+
+import java.util.Arrays;
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.opengis.filter.capability.FunctionName;
+
+public class FilterFunction_arrayAnyMatch extends FunctionExpressionImpl {
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java]: Please rename to follow java conventions, *AnyMatchArrayFunction*.

The darn "FilterFunction_" prefix was produced by a code generator back at the dawn of time.

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    public FilterFunction_arrayAnyMatch() {
+        super(NAME);
+    }
+
+    public Object evaluate(Object feature) {
+        Object[] array;
+        Object candidate;
+
+        try { // attempt to get array and perform conversion
+            array = (Object[]) getExpression(0).evaluate(feature);
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java]: if you expect an `Object[]` then ask for it rather than eating exceptions:
```
array = getExpression(0).evalulate( feature, Object[] )
```

This approach reduce the amount if error checking boilerplate and allows folks to introduce conversions, for example allowing your function to handle an array represented as a string.

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.opengis.filter.capability.FunctionName;
+
+public class FilterFunction_arrayAnyMatch extends FunctionExpressionImpl {
+    public static FunctionName NAME =
+            new FunctionNameImpl(
+                    "arrayAnyMatch",
+                    parameter(
+                            "arrayAnyMatch",
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java]: Feedback: The name "arrayAnyMatch" is still odd to me when the usual function is called `IN`, would "InArray" work for you? This would also reverse the order of the arguments:
```
InArray( candidate, array )
```

File: modules/library/main/src/test/java/org/geotools/filter/function/FilterFunction_arrayAnyMatchTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    @Test
+    public void testDoesNotMatchStrings() throws Exception {
+        FilterFactoryImpl ff = new FilterFactoryImpl();
+        Function func =
+                ff.function(
+                        "arrayAnyMatch",
+                        ff.literal(new String[] {"a", "b", "c", "d"}),
+                        ff.literal("g"));
+        assertFalse((Boolean) func.evaluate(new Object()));
+    }
```
Comment (jodygarnett) [inline modules/library/main/src/test/java/org/geotools/filter/function/FilterFunction_arrayAnyMatchTest.java]: Test the edge cases to check if type conversion is working:
```
    @Test
     public void testDoesNotMatchStrings() throws Exception {
         FilterFactoryImpl ff = new FilterFactoryImpl();
         Function func =
                 ff.function(
                         "arrayAnyMatch",
                         ff.literal("a"),
                         ff.literal("g"));
         assertFalse((Boolean) func.evaluate(new Object()));
     }
```

Can I ask how this is represented in `ECQL`?

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java [hunk 5]
```diff
@@ -930,6 +934,20 @@ private String getPrimaryKeyColumnsAsCommaSeparatedList(
         return sb.toString();
     }
 
+    public Object visit(FilterFunction_arrayAnyMatch filter, Object extraData) {
+        Expression array = getParameter(filter, 0, true);
```
Comment (jodygarnett) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java]: Defensive programming needed here to fit in with the Filter flexibility in representation of values.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java [hunk 6]
```diff
@@ -930,6 +934,20 @@ private String getPrimaryKeyColumnsAsCommaSeparatedList(
...
+        Expression candidate = getParameter(filter, 1, true);
```
Comment (jodygarnett) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java]: Do you need to check that the type of values match, will SQL be upset if you check if a boolean is in an array of strings?

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostgisPSFilterToSql.java [hunk 7]
```diff
@@ -187,6 +189,8 @@ public Object visit(PropertyIsEqualTo filter, Object extraData) {
                                     throw new RuntimeException(e);
                                 }
                             }));
+        } else if (any != null) {
+            return helper.visit(any, extraData);
```
Comment (jodygarnett) [inline modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostgisPSFilterToSql.java]: Feedback: A question that goes beyond this PR review - is the duplication necessary? Can one class extend the other.

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java (hunk 3 above)
Comment (mbarto) [reply modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_arrayAnyMatch.java]: I agree going to rename to inArray, and removing boilerplate

File: modules/library/main/src/test/java/org/geotools/filter/function/FilterFunction_arrayAnyMatchTest.java (hunk 4 above)
Comment (mbarto) [reply modules/library/main/src/test/java/org/geotools/filter/function/FilterFunction_arrayAnyMatchTest.java]: From my knowledge you cannot express array literals in ECQL, so this is only going to be used on datasource that have array type properties (we use it that way in the postgis driver)

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostgisPSFilterToSql.java (hunk 7 above)
Comment (mbarto) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/PostgisPSFilterToSql.java]: Not sure about this. I suppose the helper class is the current attempt to put some code in a common place for both.

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java (hunk 5 above)
Comment (mbarto) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java]: Can you explain more precisely what you would do here?

File: modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java (hunk 6 above)
Comment (mbarto) [reply modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java]: I am going to add checks for type matching and casting when needed

**Comment by aaime** [comment]
The appveyor failure is unrelated (download issues) but I cannot restart it, guess only @jodygarnett can
